src/sorting/merge_sort.rs

```diff
@@ -0,0 +1,124 @@
+pub fn merge_sort<T>(slice: &mut [T])
+where
+    T: Ord,
+{
+    merge_sort_cmp(slice, |a, b| a.lt(b));
+}
+
+pub fn merge_sort_cmp<T, F>(v: &mut [T], mut less: F)
+where
+    F: FnMut(&T, &T) -> bool,
+{
+    if std::mem::size_of::<T>() == 0 {
+        return;
+    }
+
+    // Create a buffer with size `v.len() * size_of::<T>()` contains uninitialized data.
+    let mut buf = Vec::with_capacity(v.len());
+
+    sort_range(v, 0, v.len(), buf.as_mut_ptr(), &mut less);
+}
+
+fn sort_range<T, F>(v: &mut [T], begin: usize, end: usize, buf: *mut T, less: &mut F)
+where
+    F: FnMut(&T, &T) -> bool,
+{
+    if end - begin <= 1 {
+        return;
+    }
+
+    let middle = (begin + end) / 2;
+
+    sort_range(v, begin, middle, buf, less);
+    sort_range(v, middle, end, buf, less);
+    merge(v, begin, middle, end, buf, less);
+}
+
+fn merge<T, F>(v: &mut [T], begin: usize, middle: usize, end: usize, buf: *mut T, mut less: F)
+where
+    F: FnMut(&T, &T) -> bool,
+{
+    struct Range {
```

Comment:
I wouldn't declare the struct in the function. Please pull out the struct declaration and have it be module private.